src/interface/report/PhaseParser.ts

```diff
@@ -0,0 +1,91 @@
+import React, { ReactNode } from 'react';
+
+import { EventType, PhaseEvent } from 'parser/core/Events';
+import { findByBossId, Phase } from 'raids';
+
+export const SELECTION_ALL_PHASES = "ALL";
+export const SELECTION_CUSTOM_PHASE = "CUSTOM";
+
+interface Props {
+  fight: {
+    // eslint-disable-next-line camelcase
+    start_time: number,
+    // eslint-disable-next-line camelcase
+    end_time: number,
+    boss: number,
+  },
+  bossPhaseEvents: PhaseEvent[],
+  children: (isLoading: boolean, phases?: {[key:string]: Phase}) => ReactNode,
+}
+
+interface State {
+  isLoading: boolean,
+  phases?: {[key:string]: Phase},
+}
+
+class PhaseParser extends React.PureComponent<Props, State> {
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      isLoading: true,
+    };
+  }
+
+  componentDidMount() {
+    // noinspection JSIgnoredPromiseFromCall
+    this.parse();
+  }
+  componentDidUpdate(prevProps: Props) {
+    const phasesChanged = this.props.bossPhaseEvents !== prevProps.bossPhaseEvents;
+    if (phasesChanged) {
+      this.setState({
+        isLoading: true,
+      });
+      // noinspection JSIgnoredPromiseFromCall
+      this.parse();
+    }
+  }
+
+  makePhases() : {[key:string]: Phase}{
+    const { bossPhaseEvents, fight } = this.props;
+    if(!bossPhaseEvents){
+      return {};
+    }
+    const phaseStarts = Array.from(new Set<string>(bossPhaseEvents.filter((e : PhaseEvent) => e.type === EventType.PhaseStart).map((e: PhaseEvent) => e.phase.key))); //distinct phase starts
```

Comment:
this is a lot of code on one line. If you're doing distinct two or more times, maybe make a function for it?

e.g. `const distinct = <T extends any>(items: T[]) => Array.from(new Set<T>(items))`